optax/contrib/_sophia.py

```diff
@@ -0,0 +1,339 @@
+# Copyright 2023 DeepMind Technologies Limited. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Sophia optimizer.
+
+A contributed implementation of the Sophia optimizer from "Sophia: A Scalable
+Stochastic Second-order Optimizer for Language Model Pre-training"
+(https://arxiv.org/abs/2305.14342) by Hong Liu, Zhiyuan Li, David Hall,
+Percy Liang, and Tengyu Ma.
+
+This contribution is heavily based on the implementation of Sophia by levanter
+(https://github.com/stanford-crfm/levanter) with some changes.
+"""
+from typing import Any, NamedTuple, Optional, Union, Callable
+from functools import partial
+
+import jax
+from jax import numpy as jnp
+from jax.random import PRNGKey
+
+from optax._src import base
+from optax._src import transform
+from optax._src.numerics import safe_int32_increment
+from optax._src.utils import canonicalize_dtype
+from optax._src.combine import chain
+from optax import tree_utils as otu
+
+
+class HutchinsonState(NamedTuple):
+  key: PRNGKey
+
+
+def hutchinson_estimator_diag_hessian(
+    random_seed: Optional[PRNGKey] = None
+):
+  """Returns a GradientTransformation that computes the diagonal of the Hessian.
+
+  The Hessian diagonal is estimated using Hutchinson's estimator, which is
+  unbiased but has high variance.
+
+  Args:
+    random_seed: Optional[PRNGKey], key used to generate random vectors.
+
+  Returns:
+    GradientTransformationExtraArgs
+  """
+
+  def init_fn(params):
+    del params
+    key = random_seed if random_seed is not None else jax.random.PRNGKey(0)
+    return HutchinsonState(key=key)
+
+  def update_fn(updates, state, params=None, obj_fn=None):
+    if params is None:
+      raise ValueError("params must be provided to hutchinson update function.")
+    if obj_fn is None:
+      raise ValueError("obj_fn must be provided to hutchinson update function.")
+    del updates
+    key, subkey = jax.random.split(state.key)
+    random_signs = otu.tree_random_like(
```

Comment:
As far as I can tell from the paper (https://arxiv.org/pdf/2305.14342, section 2.3), it computes the Hutchinson estimator using a Normal distribution, while here we use a Rademacher distribution. The Rademacher distribution should be lower variance, but perhaps it's worth to add a comment in that this deviates from what is specified in the paper?

Reply:
Hi Fabian! Yes it seemed some math worked out to lower variance with rademacher when dealing with diag hessian like in sophia but I'll actually just revert to normal for simplicity and staying true to the paper for now.